modules/broker-core/src/main/java/io/ballerina/messaging/broker/core/ObservableQueueHandlerImpl.java

```diff
@@ -0,0 +1,207 @@
+/*
+ * Copyright (c) 2018, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ *
+ * WSO2 Inc. licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.ballerina.messaging.broker.core;
+
+import io.ballerina.messaging.broker.common.ValidationException;
+import io.ballerina.messaging.broker.common.util.function.ThrowingConsumer;
+import io.ballerina.messaging.broker.core.metrics.BrokerMetricManager;
+import io.ballerina.messaging.broker.eventing.EventSync;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import javax.transaction.xa.Xid;
+
+/**
+ * Represents an Queue Handler which trigger events for the broker.
+ */
+public class ObservableQueueHandlerImpl extends QueueHandler {
+
+    private QueueHandlerImpl queueHandler;
+    private DefaultQueueHandlerEventPublisher defaultQueueHandlerEventPublisher;
+    private EventSync eventSync;
+
+    ObservableQueueHandlerImpl(Queue queue,
+                               BrokerMetricManager metricManager,
+                               EventSync eventSync,
+                               List<Integer> messageLimits) {
+        queueHandler = new QueueHandlerImpl(queue, metricManager);
```

Comment:
I assume we are using the decorator pattern here. We should decorate an implementation of QueueHandler using this class. In this line we are creating the decorated queue handler as well. I thing this is a responsibility of the factory class. Factory can create the queue handler object and pass it to the decorator class. WDYT?